SO-6477: improved lexical similarity suggester

cmark opened the pull request:
Jira: https://snowowl.atlassian.net/browse/SO-6477

cmark committed:
fix(suggest): ensure that inside the like/unlike arrays...

...we only convert true URIs to URI instances and not trying to be too
smart by converting every text into CodeSystem URI.

cmark committed:
refactor(suggest): pull up top token generation to suggestion ctx...

...from TermConceptSuggester

cmark committed:
feat(suggest): cache top tokens for the requested config to avoid...

...fetching content over and over when accessing topTokens from ctx

cmark committed:
fix(suggest): report missing concept suggester before reporting...

...any other argument issues

cmark committed:
fix(suggest): ensure that from URIs without resource type will still...

...fall back to codesystem URIs

cmark committed:
feat(suggest): implement a more sophisticated lexical similarity alg...

...based on the available capabilities of the current index mapping

cmark committed:
test(suggest): add lexical sim suggest test case and update the old ones

cmark committed:
refactor(index): use the built-in saturation function in score...

...normalization

cmark committed:
refactor(index): extract score normalization expression helper

cmark committed:
fix(core): rename argument from branch to path

cmark committed:
test(core): add JobApiAssert

cmark committed:
fix(api): report a programming related illegal arg error when...

...resourcePath is null or empty

This usually indicates that a validation rule is missing for the path
input value in the interface or somewhere in the call stack.
We could add it here, but given that this is a generic place we can only
add a generic error that might not help identify which property is
missing or erroneous.

cmark committed:
fix(core): resolve unable to search remote jobs by their type

This was a working functionality in earlier versions but due to changing
the parameters to a non-indexed format the type filter is no longer
functioning properly and causing search issues.
This commit fixes it by indexing the type parameter value from
parameters into a keyword field that can be searched.

cmark committed:
fix(core): make sure certain requests that require initialization...

...before serializing them for any reason get their initial parameters
computed

cmark committed:
fix(core): make sure we perform the same execution flow in async...

...remote job wrapper execution as in the normal flow

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/search/LexicalSimilarityTermFilter.java:
```diff
@@ -0,0 +1,222 @@
+				.withIgnoreStopwords(isIgnoreStopwords())
+				.withSynonymsEnabled(isSynonyms()),
+
+			// then all prefixes match
+			matchTextAll(fieldAlias(field, prefixFieldSuffix), getTerm())
+				.withIgnoreStopwords(isIgnoreStopwords()),
+				
+			// and last some prefixes match
+			matchTextAny(fieldAlias(field, prefixFieldSuffix), getTerm(), getMinShouldMatch())
+				.withIgnoreStopwords(isIgnoreStopwords())
+				.withSynonymsEnabled(isSynonyms())
```
I'm wondering whether prefix searches should ignore the "ignore stopwords" option since prefixes of longer words can match a stopword. Example: "android" &rarr; "and"

cmark replied:
Yes, that can cause issues, so probably yes, but I have something to showcase in person, which will cause even more issues that we need to solve. 😅

cmark committed:
fix(core): add new job doc schema revision to 2

Perform only mapping migration for this revision as we have only added a
new document field which is only required for newly initiated jobs.

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/search/LexicalSimilarityTermFilter.java:
```diff
@@ -0,0 +1,222 @@
+		private Integer minShouldMatch;
+		
+		private Boolean ignoreStopwords;
+		private boolean caseSensitive;
```
Is `caseSensitive` a mandatory option? The other two fields are `Boolean` wrappers, not primitives.

cmark replied:
Yeah, not mandatory, defaults to false. Will change the type.

cmark replied:
Done.

apeteri reviewed (approved):
LGTM with one question.

nagyo commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/search/LexicalSimilarityTermFilter.java:
```diff
@@ -0,0 +1,222 @@
+		}
+		
+		public Builder fuzzy() {
+			return fuzziness("AUTO");
```
might be worth extracting potential values to a constant class or enums?

cmark replied:
I would defer this until we have full fuzziness configuration support in this API.

nagyo reviewed (approved):
LGTM so far

cmark committed:
Merge remote-tracking branch 'origin/9.x' into feature/SO-6477-term-suggester-lexical-similarity-improvements

cmark committed:
fix(core): ensure caseSensitive setting is optional on API level

NOTE: it defauls to false in all cases meaning that all searches are performed in a case insensitive way by default.

nagyo reviewed (approved):
🍏

cmark committed:
fix(core): make rate limiting request execution optional...

...depending on whether we have the service configured or not

cmark committed:
test(jobs): add additional context info and services...

...when executing jobs in unit tests

cmark committed:
fix(auth): allow proper bundle access when versioning a resource

Fixes test failures after introducing more scoped exec context for jobs.

apeteri reviewed (approved):
🍃

nagyo reviewed (approved):
LGTM!
